File: UCF HSPT Documents/2013/Solutions/squirrels.c

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: Squirrel Terrirory (squirrels)
//
// Author:        Matt Fontaine
// Judge Data:    Mike Galletti
// C Solution:    Patrick Fenelon
// Java Solution: John Boswell
// Verifier:      SteVen Batten

#include <stdio.h>
#include <stdlib.h>

//no uniform constant in all C math libraries, thus, define it here
#define PI 3.14159265358979 

int main()
{
   int c,t;
   int *x,*y;
   int testCase, i, j;

   int min;

   //open file to read from
   FILE *in = fopen("squirrels.in", "rt");

   //read number of test cases (C)
   fscanf(in, " %d", &c);

   //iterate over test cases (campuses)
   for(testCase = 0; testCase < c; testCase++){

      //read number of trees in this test case (T)
      fscanf(in, " %d", &t);

      //allocate space for arrays to store coordinates
      x = (int*) malloc(t * sizeof(int));
      y = (int*) malloc(t * sizeof(int));

      //read in coordinates
      for(i = 0; i < t; i++){
         fscanf(in, " %d", &(x[i]));
         fscanf(in, " %d", &(y[i]));
      }

      //set to arbitrarily large value to start with
      min = 99999999; 

      //find smallest distance between trees
      for(i = 0; i < t; i++){
         for(j = 0; j < t; j++){
            if(i!=j){
               int dx = x[i] - x[j];
               int dy = y[i] - y[j];
               int distSq = dx*dx+dy*dy;
               
               //only the square of the distance is needed, 
               //this will avoid error in later calculations
               if(min > distSq) min = distSq;
            }
         }
      }

      //free memory, we're done with x and y now
      free(x);
      free(y);

      //compute answer, keep in mind we have the 2*r^2 stored in min, 
      //to account for this multiply by PI and divide by 4 to get the area
      double answer = PI / 4.0 * min; 

      //print answer
      printf("Campus #%d:\n", testCase+1);
      printf("Maximum territory area = %.3f\n\n", answer);

   }   

   //no error happened so return 0
   fclose(in);
   return 0;
}

File: UCF HSPT Team/src/y2013/squirrels.java

// 27th Annual UCF High School Programming Tournament
//
// Problem Name: Squirrel Terrirory (squirrels)
//
// Author:        Matt Fontaine
// Judge Data:    Mike Galletti
// C Solution:    Patrick Fenelon
// Java Solution: John Boswell
// Verifier:      SteVen Batten

import java.util.*;
import java.io.*;

public class squirrels {

   //Infinity (a number larger than we will be using).
   static final int oo = (int)1e9;
   
   //Calculate the magnitude of x/y differences (simplifies the code).
   static double magnitude(int x, int y) {
      return Math.sqrt(x*x+y*y);
   }

   public static void main(String[] args) throws IOException {
      //Initialize scanner.
      Scanner in = new Scanner(new File("squirrels.in"));

      //Read in the number of campuses.
      int campuses = in.nextInt();
      
      //For each campus...
      for (int c=0; c<campuses; c++) {

         //Read in the number of trees.
         int trees = in.nextInt();

         //Initialize arrays for X/Y values.
         int[] X = new int[trees], Y = new int[trees];

         //For each tree, read in its location.
         for (int t=0; t<trees; t++) {
            //Read in the X and Y values of that tree.
            X[t] = in.nextInt();
            Y[t] = in.nextInt();
         }

         //Initialize the max radius so something huge, so that it is 
         //guaranteed larger than the actual minimum.
         double radius = oo;

         //For each pair of trees...
         for (int a=0; a<trees; a++)
            for (int b=a+1; b<trees; b++)
               //Calculate the distance between the trees. This divided by two
               // is the most equal (greatest for both) split of circle radii.
               // The best radius must be less than or equal to this value.
               radius = Math.min(radius,magnitude(X[a]-X[b],Y[a]-Y[b])/2.0);

         //Output the answer.
         System.out.printf("Campus #%d:\n",c+1);
         System.out.printf("Maximum territory area = %.3f\n",
                           radius*radius*Math.PI);
         System.out.println();
      }

      // Close file
      in.close();
   }
}
